File: app/src/main/java/github/leavesc/wififiletransfer/FileSenderActivity.java
package github.leavesc.wififiletransfer;

import android.app.ProgressDialog;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import java.io.File;
import java.text.MessageFormat;

import github.leavesc.wififiletransfer.common.Constants;
import github.leavesc.wififiletransfer.manager.WifiLManager;
import github.leavesc.wififiletransfer.model.FileTransfer;
import github.leavesc.wififiletransfer.service.FileSenderService;

/**
 * @Author: leavesC
 * @Date: 2018/4/3 14:53
 * @Desc:
 * @Github：https://github.com/leavesC
 */
public class FileSenderActivity extends BaseActivity {

    public static final String TAG = "FileSenderActivity";

    private static final int CODE_CHOOSE_FILE = 100;

    private FileSenderService fileSenderService;

    private ProgressDialog progressDialog;

    private final FileSenderService.OnSendProgressChangListener progressChangListener = new FileSenderService.OnSendProgressChangListener() {

        @Override
        public void onStartComputeMD5() {
            runOnUiThread(() -> {
                if (isCreated()) {
                    progressDialog.setTitle("发送文件");
                    progressDialog.setMessage("正在计算文件的MD5码");
                    progressDialog.setMax(100);
                    progressDialog.setProgress(0);
                    progressDialog.setCancelable(false);
                    progressDialog.show();
                }
            });
        }

        @Override
        public void onProgressChanged(final FileTransfer fileTransfer, final long totalTime, final int progress, final double instantSpeed, final long instantRemainingTime, final double averageSpeed, final long averageRemainingTime) {
            runOnUiThread(() -> {
                if (isCreated()) {
                    progressDialog.setTitle("正在发送文件： " + new File(fileTransfer.getFilePath()).getName());
                    if (progress != 100) {
                        progressDialog.setMessage("文件的MD5码：" + fileTransfer.getMd5()
                                + "\n\n" + "总的传输时间：" + totalTime + " 秒"
                                + "\n\n" + "瞬时-传输速率：" + (int) instantSpeed + " Kb/s"
                                + "\n" + "瞬时-预估的剩余完成时间：" + instantRemainingTime + " 秒"
                                + "\n\n" + "平均-传输速率：" + (int) averageSpeed + " Kb/s"
                                + "\n" + "平均-预估的剩余完成时间：" + averageRemainingTime + " 秒"
                        );
                    }
                    progressDialog.setProgress(progress);
                    progressDialog.setCancelable(true);
                    progressDialog.show();
                }
            });
        }

        @Override
        public void onTransferSucceed(FileTransfer fileTransfer) {
            runOnUiThread(() -> {
                if (isCreated()) {
                    progressDialog.setTitle("文件发送成功");
                    progressDialog.setCancelable(true);
                    progressDialog.show();
                }
            });
        }

        @Override
        public void onTransferFailed(FileTransfer fileTransfer, final Exception e) {
            runOnUiThread(() -> {
                if (isCreated()) {
                    progressDialog.setTitle("文件发送失败");
                    progressDialog.setMessage("异常信息： " + e.getMessage());
                    progressDialog.setCancelable(true);
                    progressDialog.show();
                }
            });
        }
    };

    private final ServiceConnection serviceConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            FileSenderService.MyBinder binder = (FileSenderService.MyBinder) service;
            fileSenderService = binder.getService();
            fileSenderService.setProgressChangListener(progressChangListener);
            Log.e(TAG, "onServiceConnected");
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            fileSenderService = null;
            bindService(FileSenderService.class, serviceConnection);
            Log.e(TAG, "onServiceDisconnected");
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_file_sender);
        initView();
        bindService(FileSenderService.class, serviceConnection);
    }

    private void initView() {
        setTitle("发送文件");
        TextView tv_hint = findViewById(R.id.tv_hint);
        tv_hint.setText(MessageFormat.format("在发送文件前需要先连上文件接收端开启的Wifi热点\n热点名：{0} \n密码：{1}", Constants.AP_SSID, Constants.AP_PASSWORD));
        progressDialog = new ProgressDialog(this);
        progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        progressDialog.setCancelable(false);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.setTitle("发送文件");
        progressDialog.setMax(100);
        progressDialog.setIndeterminate(false);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (fileSenderService != null) {
            fileSenderService.setProgressChangListener(null);
            unbindService(serviceConnection);
        }
        progressDialog.dismiss();
    }

    public void sendFile(View view) {
        if (!Constants.AP_SSID.equals(WifiLManager.getConnectedSSID(this))) {
            showToast("当前连接的 Wifi 并非文件接收端开启的 Wifi 热点，请重试或者检查权限");
            return;
        }
        navToChosePicture();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == CODE_CHOOSE_FILE) {
            if (resultCode == RESULT_OK) {
                String imageUri = data.getData().toString();
                Log.e(TAG, "文件路径：" + imageUri);
                FileSenderService.startActionTransfer(this, imageUri,
                        WifiLManager.getHotspotIpAddress(this));
            }
        }
    }

    private void navToChosePicture() {
        Intent intent = new Intent(Intent.ACTION_PICK, null);
        intent.setDataAndType(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, "image/*");
        startActivityForResult(intent, CODE_CHOOSE_FILE);
    }

}

File: app/src/main/java/github/leavesczy/wififiletransfer/ui/FileSenderActivity.kt
package github.leavesczy.wififiletransfer.ui

import android.app.ProgressDialog
import android.content.ComponentName
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.widget.Button
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import github.leavesczy.wififiletransfer.R
import github.leavesczy.wififiletransfer.service.Constants
import github.leavesczy.wififiletransfer.service.FileSenderService
import github.leavesczy.wififiletransfer.service.FileSenderService.OnSendProgressChangListener
import github.leavesczy.wififiletransfer.service.FileTransfer
import github.leavesczy.wififiletransfer.utils.WifiLManager
import java.io.File
import java.text.MessageFormat

/**
 * @Author: leavesCZY
 * @Desc:
 */
class FileSenderActivity : BaseActivity() {

    private val getContentLaunch = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { imageUri ->
        if (imageUri != null) {
            log("文件路径： $imageUri")
            FileSenderService.startActionTransfer(
                this, imageUri.toString(),
                WifiLManager.getHotspotIpAddress(this)
            )
        }
    }

    private var fileSenderService: FileSenderService? = null

    private val progressChangListener: OnSendProgressChangListener =
        object : OnSendProgressChangListener {
            override fun onStartComputeMD5() {
                runOnUiThread {
                    if (isCreated) {
                        progressDialog.setTitle("发送文件")
                        progressDialog.setMessage("正在计算文件的MD5码")
                        progressDialog.max = 100
                        progressDialog.progress = 0
                        progressDialog.setCancelable(false)
                        progressDialog.show()
                    }
                }
            }

            override fun onProgressChanged(
                fileTransfer: FileTransfer,
                totalTime: Long,
                progress: Int,
                instantSpeed: Double,
                instantRemainingTime: Long,
                averageSpeed: Double,
                averageRemainingTime: Long
            ) {
                runOnUiThread {
                    if (isCreated) {
                        progressDialog.setTitle("正在发送文件： " + File(fileTransfer.filePath).name)
                        if (progress != 100) {
                            progressDialog.setMessage(
                                "文件的MD5码：${fileTransfer.md5}\n" +
                                        "总的传输时间：$totalTime 秒\n" +
                                        "瞬时-传输速率：${instantSpeed.toInt()} Kb/s\n" +
                                        "瞬时-预估的剩余完成时间：$instantRemainingTime 秒\n" +
                                        "平均-传输速率：${averageSpeed.toInt()} Kb/s\n" +
                                        "平均-预估的剩余完成时间：$averageRemainingTime 秒"
                            )
                        }
                        progressDialog.progress = progress
                        progressDialog.setCancelable(true)
                        progressDialog.show()
                    }
                }
            }

            override fun onTransferSucceed(fileTransfer: FileTransfer?) {
                runOnUiThread {
                    if (isCreated) {
                        progressDialog.setTitle("文件发送成功")
                        progressDialog.setCancelable(true)
                        progressDialog.show()
                    }
                }
            }

            override fun onTransferFailed(fileTransfer: FileTransfer?, e: Exception) {
                runOnUiThread {
                    if (isCreated) {
                        progressDialog.setTitle("文件发送失败")
                        progressDialog.setMessage("异常信息： " + e.message)
                        progressDialog.setCancelable(true)
                        progressDialog.show()
                    }
                }
            }
        }
    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val binder = service as FileSenderService.MyBinder
            fileSenderService = binder.service
            fileSenderService?.setProgressChangListener(progressChangListener)
            log("onServiceConnected")
        }

        override fun onServiceDisconnected(name: ComponentName) {
            fileSenderService = null
            bindService(FileSenderService::class.java, this)
            log("onServiceDisconnected")
        }
    }

    private val tvHint by lazy {
        findViewById<TextView>(R.id.tvHint)
    }

    private val sendFile by lazy {
        findViewById<Button>(R.id.sendFile)
    }

    private lateinit var progressDialog: ProgressDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_file_sender)
        initView()
        bindService(FileSenderService::class.java, serviceConnection)
    }

    private fun initView() {
        setTitle("发送文件")
        tvHint.text = MessageFormat.format(
            "在发送文件前需要先连上文件接收端开启的Wifi热点\n热点名：{0} \n密码：{1}",
            Constants.AP_SSID,
            Constants.AP_PASSWORD
        )
        sendFile.setOnClickListener {
            if (Constants.AP_SSID != WifiLManager.getConnectedSSID(this)) {
                showToast("当前连接的 Wifi 并非文件接收端开启的 Wifi 热点，请重试或者检查权限")
                return@setOnClickListener
            }
            getContentLaunch.launch("image/*")
        }
        progressDialog = ProgressDialog(this).apply {
            setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
            isIndeterminate = false
            setCancelable(false)
            setCanceledOnTouchOutside(false)
            max = 100
            setTitle("发送文件")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (fileSenderService != null) {
            fileSenderService?.setProgressChangListener(null)
            unbindService(serviceConnection)
        }
        progressDialog.dismiss()
    }

}